File: GO/Programmers/Hash/베스트앨범.go
/*
https://programmers.co.kr/learn/courses/30/lessons/42579

풀이 :
*/

package main

import (
	"fmt"
)

func main() {
	genres := []string{"classic", "pop", "classic", "classic", "pop"}
	plays := []int{500, 600, 150, 800, 2500}
	fmt.Println("결과 : ", solution(genres, plays))

}

// 맵으로 클래식 3000, 팝송 1500 해서 숫자 큰 거 대로 정렬한 뒤 또 돌려서 거기서 큰 거 두개씩만 갖고오기
// 클래식, 팝송, 발라드 나온다치면 클래식 2개 넣고 팝송 2개 넣고 발라드 2개 넣으면될듯

func solution(genres []string, plays []int) []int {
	genrePlays := make(map[string]int)
	genreInterface := []interface{}{}
	// genrePlays : [classic] = 1450 , [pop] = 3100 이런 식으로 되게끔 넣음
	for i := 0; i < len(genres); i++ {
		if 0 < genrePlays[genres[i]] {
			genrePlays[genres[i]] += plays[i]
		} else {
			genrePlays[genres[i]] = plays[i]
		}
	}
	for i, v := range genrePlays {
		genreInterface = append(genreInterface, i, v)
	}
	for i := 1; i < len(genreInterface)-2; i += 2 {
		if genreInterface[i+2] != nil {
			if genreInterface[i].(int) < genreInterface[i+2].(int) {
				genreInterface[i-1], genreInterface[i], genreInterface[i+1], genreInterface[i+2] =
					genreInterface[i+1], genreInterface[i+2], genreInterface[i-1], genreInterface[i]
			}
		}
	}
	fmt.Println("genreInterface : ", genreInterface)
	return plays
}

File: PYTHON/Programmers/해시/베스트앨범.py
'''
https://programmers.co.kr/learn/courses/30/lessons/42579

풀이 : all_kind에 set을 통해 장르의 종류를 도출합니다. 그리고 장르의 종류를 순회하며 genre_songs 에 장르의 종류를 Key로, 총 플레이 횟수와 모든 [노래 번호, 플레이 횟수]를 Value로 저장합니다.
    그 후 ranking_songs에 총 플레이 횟수가 높은 순서대로 저장하고 all_songs에 모든 [노래 번호, 플레이 횟수]를 랭킹이 높은 장르 순으로 리스트로 구분하여 저장합니다.
    그리고 장르 별 모든 노래를 노래의 플레이 횟수 순서로 정렬한 뒤 album에 추가합니다. 만약 장르 별 노래 수가 2개 이상이라면 2개를 추가합니다.
'''

from collections import deque

def songs_by_genre_ranking(genres, plays):
    all_kind = list(set(genres))
    genre_songs = {}
    while all_kind:
        kind = all_kind.pop()
        genre_songs[kind] = [0]
        for idx, (genre, play) in enumerate(zip(genres, plays)):
            if genre == kind:
                genre_songs[kind][0] += play
                genre_songs[kind].append([idx, play])
    ranking_songs = sorted(genre_songs.values(), key=lambda x:x[0], reverse=True)
    all_songs = []
    for i in ranking_songs:
        all_songs.append(i[1:])
    return all_songs

def pick_best_song(album, songs):
    songs_ranking = sorted(songs, key=lambda x:x[1], reverse=True)
    album.append(songs_ranking[0][0])
    if len(songs_ranking) >= 2:
        album.append(songs_ranking[1][0]) 
    
def solution(genres, plays):
    album = []
    all_songs = songs_by_genre_ranking(genres,plays)
    for songs in all_songs:
        pick_best_song(album, songs)
    return album